src/modules/cpu/kernel/box_filter.hpp

```diff
@@ -0,0 +1,2314 @@
+    pxLower2 = _mm_mulhi_epi16(pxLower2, pxConvolutionFactor);
+    pxUpper1 = _mm_mulhi_epi16(pxUpper1, pxConvolutionFactor);
+
+    // saturate 16 bit values to 8 bit values and store in resultant registers
+    pxDst[0] = _mm_packus_epi16(pxLower1, pxLower2);
+    pxDst[1] = _mm_packus_epi16(pxUpper1, xmm_px0);
+}
+
+// -------------------- 7x7 kernel size - U8 bitdepth compute functions --------------------
+
+inline void compute_box_filter_u8_u8_7x7_24_host_pln(__m256i *pxRow, __m128i *pxDst, const __m128i &pxConvolutionFactor)
+{
+    // pack lower half of each of 7 loaded row values from 8 bit to 16 bit and add
+    __m256i pxLower, pxUpper;
+    pxLower = _mm256_unpacklo_epi8(pxRow[0], avx_px0);
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[1], avx_px0));
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[2], avx_px0));
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[3], avx_px0));
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[4], avx_px0));
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[5], avx_px0));
+    pxLower = _mm256_add_epi16(pxLower, _mm256_unpacklo_epi8(pxRow[6], avx_px0));
+
+    // pack higher half of each of 7 loaded row values from 8 bit to 16 bit and add
+    pxUpper = _mm256_unpackhi_epi8(pxRow[0], avx_px0);
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[1], avx_px0));
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[2], avx_px0));
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[3], avx_px0));
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[4], avx_px0));
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[5], avx_px0));
+    pxUpper = _mm256_add_epi16(pxUpper, _mm256_unpackhi_epi8(pxRow[6], avx_px0));
+
+    // get 4 SSE registers from above 2 AVX registers to arrange as per required order
+    __m128i pxLower1, pxLower2, pxUpper1, pxUpper2;
+    pxLower1 = _mm256_castsi256_si128(pxLower);
+    pxLower2 = _mm256_castsi256_si128(pxUpper);
+    pxUpper1 = _mm256_extracti128_si256(pxLower, 1);
+    pxUpper2 = _mm256_extracti128_si256(pxUpper, 1);
+
+    // get the final accumalated result for first 8 elements
+    __m128i pxTemp[6];
+    pxTemp[0] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 1), xmm_pxMaskRotate0To1);
+    pxTemp[1] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 3), xmm_pxMaskRotate0To3);
+    pxTemp[2] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 7), xmm_pxMaskRotate0To5);
+    pxTemp[3] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 15), xmm_pxMaskRotate0To7);
+    pxTemp[4] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 31), xmm_pxMaskRotate0To9);
+    pxTemp[5] = _mm_shuffle_epi8(_mm_blend_epi16(pxLower1, pxLower2, 63), xmm_pxMaskRotate0To11);
+    pxLower1 = _mm_add_epi16(pxLower1, pxTemp[0]);
+    pxLower1 = _mm_add_epi16(pxLower1, pxTemp[1]);
```

Comment:
Its probably fine for 3x3, but for 7x7 and 9x9 the _mm_add_epi16() increases to 18 and 24.
Since pxTemp is a simple array and register locations in memory will be contiguous, after the shuffles, you can try one of these 3 tricks and check if any gives better performance:
```
1) _mm256_setr_m128i() // looks like this option is higher on latency
2) A AVX2 cast - _mm256_add_epi16(_mm256_castsi128_si256(pxTemp[0]), _mm256_castsi128_si256(pxTemp[2])) // trying to add 0,1 to 2,3 of pxTemp in avx2.
3) A C++ reinterpret pointer cast - _mm256_add_epi16(*reinterpret_cast<__m256i *>(&pxTemp[0]), *reinterpret_cast<__m256i *>(&pxTemp[2])) // trying to again add 0,1 to 2,3 of pxTemp in avx2.
```

Reply:
With this approach, we can add pxTemp[0] and pxTemp[1] with pxTemp[2] and pxTemp[3], but we also need to add pxTemp[0] with pxTemp[1] and pxTemp[2] with pxTemp[3] and also pxTemp[1] with   pxTemp[2], pxTemp[0] with pxTemp[3]